### ISSUE TITLE
Lagging/Stuterring during playback

### ISSUE THREAD

**Issue by Lazmarr** [garbear/kodi-steamlink#10]
Playback lags and stutters. I have confirmed it is only with the Steam Link.
Tried the same files on the Firestick/Firebox and no issues with playback.
Has anyone got a solution for this?

**Comment by solenum** [comment]
Hi,

I've just installed this and have come across the same issues, I get a *lot* of artifacting and stuff, looks like the bitrate is ass but the media plays fine on other devices.

I'm getting the media from a SMB/samba server running locally, if I play the media on another device/computer (still from the SMB share) it runs perfectly fine.

Specifically, its like parts of the video simply arent updating, often it will go from a light to dark scene and the video will appear a little brighter than it should and then it will update to look *correct* to some extent.

This is with a large range of media types, with mp4 I get this artifacting, with mkv I seem to get a lot of audio de-sync, its pretty rough.

**Comment by whise** [comment]
I notice that if you set kodi buffer cache to zero you get much less lag and audio de-sync....

**Comment by henkall** [comment]
how did you do that? @whise

**Comment by whise** [comment]
Install the "maintenance addon" it's available in any number of sites.

**Comment by solenum** [comment]
@whise any way to do this *without* a plugin?

**Comment by whise** [comment]
Yeah sure if you can access the filesystem. 

Here is a guide, but there is plenty more on kodi wiki and google

http://androidpcreview.com/fix-kodi-buffering-problems-heres-how-to-modify-the-kodi-cache/2611/

**Comment by savely-krasovsky** [comment]
Well, is there some solution? Kodi on Steam Link literally still unusable. All videos lagging and shuttering.

**Comment by solenum** [comment]
I never found a solution and gave up on using this until the issues are fixed.

The lagging/stuttering/artifacting was just to much to make viewing videos enjoyable.

**Comment by SNWRbxT** [comment]
Bought one of these on the steam sale.

After failing to find any rs232 points on the board with my limited EE skills I decided to give the native build a chance!

Had the same video and audio playback issues mentioned in this thread.

What I did:

- Set the interface of the steam link to 720p (1080p maybe works too but assumed it would be less demanding) this in turn forces kodi to 720
- After RTFM here: http://kodi.wiki/view/HOW-TO:Modify_the_video_cache I came up with the following config:

`<advancedsettings>`
`<cache>`
`  <buffermode>1</buffermode>`
`  <memorysize>20971520</memorysize>`
`  <readfactor>20</readfactor>`
`</cache>`
`</advancedsettings>`


- The cache settings fixed the video lag, and I would say it is perfect (tested >4GB x264 files over smb). But I was using a USB audio device for 5.1 and was still getting audio sync issues. After disabling this and switching back to audio over HDMI from the steam link settings I had acceptable audio and video for a $5 device.

Also - regarding key based SSH access on the device:

> https://blogger.davidmanouchehri.com/2017/12/steam-link-security-truncated-password.html
> DES being used for BusyBox, and truncates passwords to have a length of only the first 8 characters.
> As a temporary fix (provided by Evan Powers), users can manually force the root account to use SHA512 instead.
> passwd -a sha512 root

/etc/ssh/sshd_config:

PubkeyAuthentication yes           
AuthorizedKeysFile      .ssh/authorized_keys

after the usual steps you will need to fix the permissions for ssh to be happy:
`chmod 755 /home/steam/`
`chmod 700 /home/steam/.ssh`
`chmod 600 /home/steam/.ssh/authorized_keys`

**YMMV**

Thanks garbear for taking the time to maintain this build.

**Comment by peterstory** [comment]
@SNWRbxT Thanks, that change fixed my playback problems, too! I was also streaming over SMB. 

For reference, I created the file here: `/home/apps/kodi/.home/.kodi/userdata/advancedsettings.xml`

**Comment by ghost** [comment]
@peterstory This fixed the stuttering video after pauses?

**Comment by peterstory** [comment]
@AMDSY3D I initially thought it did, but now I'm not so sure. A potentially confounding factor was that my PS3 controller was causing small amount of scrubbing, randomly. It seems that the analog sticks were so sensitive that they were randomly causing the scrubbing, even when I wasn't touching the controller. I fixed that issue by configuring the directional pad to be used for scrubbing, instead of the analog stick.

**Comment by garbear** [comment]
Kodi has a deadzone setting. It's well hidden - visit Settings -> System settings -> Input -> Peripherals.

Kodi v18 contains a special feature (the ability to play ROMs alongside videos) so I'll make the deadzone setting more prominent in our next release.

**Comment by forid786** [comment]
I've personally picked up a Steam Link in the sale, just to run videos and media off my FreeNas box using kodi. I'm pretty sure others have bought the Link to run Kodi as well. 

Kodi on the Steam link as a whole seems quite laggy and jumpy at times. Is this mainly caused by the hardware or software? I would've thought that the stream link would be good enough to run kodi.

**Comment by garbear** [comment]
The Steam Link is largely constrained by the weak CPU. Kodi's menu system is quite CPU-heavy, very few triangles but a very large amount of skinning conditions to calculate every frame. If you want less menu lag, a light-weigh skin will help (not sure if we have a light-weight one in our repo, usually skinners overload their babies with features).

Because it's accelerated, HD video should be smoother than the menus.

**Comment by forid786** [comment]
I've tried a few light weight themes, the experience has been better. But it's really slow compared to running Kodi on PC or on my Android phone. When streaming from SMB, I've had jumps in videos, artefact and subtitles being too fast or too slow.

I think for the time being, I might go with running Kodi on the PC route and leave the the steam link with what it's good at. Streaming. But it's great to see you working on this native Kodi addon.

**Comment by StripedMonkey** [comment]
Trying what SNWRbxT said helped out a ton, to the point where its almost watchable. I refuse to believe that it can't handle HD/1080p, and yet,  that is what it seems to struggle with. Are there any performance optimizations that can be done beyond whats already been suggested?

**Comment by jonsanta** [comment]
> Trying what SNWRbxT said helped out a ton, to the point where its almost watchable. I refuse to believe that it can't handle HD/1080p, and yet, that is what it seems to struggle with. Are there any performance optimizations that can be done beyond whats already been suggested?

did really help? I tryed with that file, also with maintenance addon, seems like buffer size cant be changed on the build i was using. wich build did u use?

**Comment by StripedMonkey** [comment]
Originally I used Kodi V18 on steam link build 683 but couldn't get it to work so I nuked /home/apps/kodi and installed V17.6. After adding the startup script for NTP (See #28 ) and adding the little blurb SN wrote to `/home/apps/kodi/.home/.kodi/userdata/advancedsettings.xml` I restarted and got a dramatically higher performance out of Kodi but not great. Every few seconds the video would freeze, it was still unwatchable. After about 10 minutes it seemed to sort itself out, and while I think I might still see some hitches its much better than it was before. Like I said, its still not great performance, especially from a device that should be perfectly suitable for 1080p streaming at a good frame rate.

**Comment by shipilovr** [comment]
Does anyone have audio and subs resync problem after pausing or moving back/forward video? If it starts from begining everything is OK. I checked CPU usage by top-command over SSH within this moment and it is 100%. Seems SL can't just work out loading on it...

**Comment by StripedMonkey** [comment]
Yup. Pausing causes the issue for me, but Fast forwarding/backwards doesn't *always* cause it.

**Comment by MadManMoon** [comment]
Sorry to awaken an old thread, I'm still seeing this issue. So I'm adding my comment to indicate that and also to show my troubleshooting.

I'm going to try the work arounds that are in this thread, but perhaps this is widespread enough to ensure that they are included in the next release? (_that's not an insistent thing, just thinking aloud_)

Did I also read somewhere that it's basically pointless to try any files that are not `264`, due to the on chip processing? Was that this thread?

Anyway, like I say, I'm going to try to downgrade some day soon, so I'll see if I can just find a way to include the workarounds in the version that I install, rather than install **then** put any workaround in. 👍

**Comment by garbear** [comment]
Yes it's true that we're limited to h264, it's the only format that the SL hardware-accelerates because it's the only format needed for game streaming.

I completed the port to Kodi v18 Leia, but the performance suffered even more due to increased processing demands of the newer version. As such, I've given up on updating the SL to newer versions.

The v17 release here never saw fixes for all A/V sync issues either. The SL is geared toward real-time streaming where A/V sync isn't an issue (you just want to show everything as fast as possible). You also don't have B-frames in realtime, because going "B"idirectional in time adds latency. I manually enable a B-frame mode for Kodi, but A/V sync then becomes a guessing game that I didn't play very well.

**Comment by StripedMonkey** [comment]
Perhaps this isn't the thread to ask it but I'd like to know if anyone has a suggestion for a "practical" use of it/a good alternative to this in order to do from-the-couch movie watching. Specifically for the SL ofc. I want to put it to use in some way.
